File: nil/cmd/nil/internal/smartaccount/new.go
package smartaccount

import (
	"fmt"

	"github.com/NilFoundation/nil/nil/cmd/nil/common"
	"github.com/NilFoundation/nil/nil/cmd/nil/internal/config"
	"github.com/NilFoundation/nil/nil/common/check"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/NilFoundation/nil/nil/services/cliservice"
	"github.com/spf13/cobra"
)

var defaultNewSmartAccountAmount = types.NewValueFromUint64(100_000_000)

func NewCommand(cfg *common.Config) *cobra.Command {
	serverCmd := &cobra.Command{
		Use:   "new",
		Short: "Create a new smart account with some initial balance on the cluster",
		RunE: func(cmd *cobra.Command, args []string) error {
			return runNew(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	setFlags(serverCmd)

	return serverCmd
}

func setFlags(cmd *cobra.Command) {
	params.salt = *types.NewUint256(0)
	cmd.Flags().Var(
		&params.salt,
		saltFlag,
		"The salt for the smart account address calculation")

	cmd.Flags().Var(
		types.NewShardId(&params.shardId, types.BaseShardId),
		shardIdFlag,
		"Specify the shard ID to interact with",
	)

	cmd.Flags().Var(
		&params.FeeCredit,
		feeCreditFlag,
		"The fee credit for smart account creation. If set to 0, it will be estimated automatically",
	)

	params.newSmartAccountAmount = defaultNewSmartAccountAmount
	cmd.Flags().Var(
		&params.newSmartAccountAmount,
		amountFlag,
		"The initial balance (capped at 10'000'000). The deployment fee will be subtracted from this balance",
	)
}

func runNew(cmd *cobra.Command, _ []string, cfg *common.Config) error {
	amount := params.newSmartAccountAmount
	if amount.Cmp(defaultNewSmartAccountAmount) > 0 {
		logger.Warn().
			Msgf("The specified balance (%s) is greater than the limit (%s). Decrease it.", &params.newSmartAccountAmount, defaultNewSmartAccountAmount)
		amount = defaultNewSmartAccountAmount
	}

	faucet, err := common.GetFaucetRpcClient()
	if err != nil {
		return err
	}
	srv := cliservice.NewService(cmd.Context(), common.GetRpcClient(), cfg.PrivateKey, faucet)
	check.PanicIfNotf(cfg.PrivateKey != nil, "A private key is not set in the config file")
	smartAccountAddress, err := srv.CreateSmartAccount(params.shardId, &params.salt, amount, params.FeeCredit, &cfg.PrivateKey.PublicKey)
	if err != nil {
		return err
	}

	if err := config.PatchConfig(map[string]interface{}{
		config.AddressField: smartAccountAddress.Hex(),
	}, false); err != nil {
		logger.Error().Err(err).Msg("failed to update the smart account address in the config file")
	}

	if !common.Quiet {
		fmt.Print("New smart account address: ")
	}
	fmt.Println(smartAccountAddress.Hex())
	return nil
}

File: clijs/src/commands/smart-account/new.ts
import { type Hex, LocalECDSAKeySigner, SmartAccountV1, bytesToHex } from "@nilfoundation/niljs";
import { Flags } from "@oclif/core";
import { BaseCommand } from "../../base.js";
import { ConfigKeys } from "../../common/config.js";
import { logger } from "../../logger.js";
import { bigintFlag } from "../../types";

export const DefualtNewSmartAccountAmount = 1_000_000_000_000_000n;

export default class SmartAccountNew extends BaseCommand {
  static override description = "Create a new smart account";

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  static flags = {
    salt: Flags.integer({
      char: "s",
      description: "The salt for the smart account address calculation",
      required: false,
      default: 0,
    }),
    shardId: Flags.integer({
      char: "i",
      description: "Specify the shard ID(>= 1) to interact with",
      required: false,
      default: 1,
      parse: async (input) => {
        const parsed = Number.parseInt(input, 10);
        if (Number.isNaN(parsed)) {
          throw new Error("Shard ID must be a number");
        }
        if (parsed < 1) {
          throw new Error("Shard ID must be greater than or equal to 1");
        }
        return parsed;
      },
    }),
    feeCredit: bigintFlag({
      char: "f",
      description:
        "The fee credit for smart account creation. If set to 0, it will be estimated automatically",
      required: false,
    }),
    amount: bigintFlag({
      char: "a",
      description:
        "The initial balance (capped at 100'000'000). The deployment fee will be subtracted from this balance",
      required: false,
      default: DefualtNewSmartAccountAmount,
    }),
  };

  public async run(): Promise<Hex> {
    const { flags } = await this.parse(SmartAccountNew);

    if (flags.amount > DefualtNewSmartAccountAmount) {
      logger.warn(
        `The specified balance (${flags.amount}) is greater than the limit (${DefualtNewSmartAccountAmount}). The default value is used.`,
      );
      flags.amount = DefualtNewSmartAccountAmount;
    }

    const privateKey = this.cfg?.[ConfigKeys.PrivateKey];
    if (!privateKey) {
      throw new Error(
        "Private key not found in config. Perhaps you need to run 'keygen new' first?",
      );
    }

    const signer = new LocalECDSAKeySigner({
      privateKey: privateKey as Hex,
    });

    if (!this.rpcClient) {
      throw new Error("RPC client is not initialized");
    }
    if (!this.faucetClient) {
      throw new Error("Faucet client is not initialized");
    }

    const pubkey = signer.getPublicKey();
    const smartAccount = new SmartAccountV1({
      pubkey: pubkey,
      salt: BigInt(flags.salt),
      shardId: flags.shardId,
      client: this.rpcClient,
      signer,
    });
    const smartAccountAddress = smartAccount.address;

    logger.debug(`Withdrawing ${flags.amount} to ${smartAccountAddress}`);

    const faucets = await this.faucetClient.getAllFaucets();
    await this.faucetClient.topUpAndWaitUntilCompletion(
      {
        amount: flags.amount,
        smartAccountAddress: smartAccountAddress,
        faucetAddress: faucets.NIL,
      },
      this.rpcClient,
    );

    this.info("Deploying smart account...");
    const tx = await smartAccount.selfDeploy(true);
    this.info(`Successfully deployed smart account with tx hash: ${bytesToHex(tx)}`);

    if (this.quiet) {
      this.log(smartAccountAddress);
    } else {
      this.log(`SmartAccount address: ${smartAccountAddress}`);
    }
    this.configManager?.updateConfig(
      ConfigKeys.NilSection,
      ConfigKeys.Address,
      smartAccountAddress,
    );
    return smartAccountAddress;
  }
}
